src/modules/job-exec/exec.c

```diff
@@ -56,32 +58,43 @@ struct exec_ctx {
     struct jobinfo *job;
 
     const char * mock_exception;   /* fake exception */
+    struct idset *barrier_pending_ranks;
     int barrier_enter_count;
     int barrier_completion_count;
     int exit_count;
+
+    double barrier_timeout;
+    flux_watcher_t *shell_barrier_timer;
 };
 
 static void exec_ctx_destroy (struct exec_ctx *tc)
 {
     if (tc) {
         int saved_errno = errno;
+        idset_destroy (tc->barrier_pending_ranks);
+        flux_watcher_destroy (tc->shell_barrier_timer);
         free (tc);
         errno = saved_errno;
     }
 }
 
-static struct exec_ctx *exec_ctx_create (json_t *jobspec)
+static struct exec_ctx *exec_ctx_create (json_t *jobspec,
+                                         const struct idset *ranks)
 {
     struct exec_ctx *ctx = calloc (1, sizeof (*ctx));
-    if (ctx == NULL)
+    if (!ctx || !(ctx->barrier_pending_ranks = idset_copy (ranks))) {
+        exec_ctx_destroy (ctx);
         return NULL;
+    }
+    ctx->barrier_timeout = 1800.;
     (void) json_unpack (jobspec,
-                        "{s:{s:{s:{s:{s:s}}}}}",
+                        "{s:{s:{s:{s:{s?s s?F}}}}}",
                         "attributes",
                           "system",
                             "exec",
                               "bulkexec",
-                                "mock_exception", &ctx->mock_exception);
+                                "mock_exception", &ctx->mock_exception,
+                                "barrier-timeout", &ctx->barrier_timeout);
```

Comment:
Suggestion: make everything optional, add `!` at the end of the `bulkexec` object to indicate extra keys are not allowed, and then check the return value of `json_unpack()` and fail on error.  That way config typos in test scripts (misspelled key, wrong type) will be a more obvious failure.

That's not an issue with this PR though - feel free to ignore.

Reply:
Good suggestion! I'll tack on a couple commits to this PR.